core/src/apps/cardano/layout.py

```diff
@@ -332,6 +362,65 @@ async def show_warning_tx_output_contains_tokens(ctx: wire.Context) -> None:
     )
 
 
+async def show_warning_tx_output_contains_datum_hash(
+    ctx: wire.Context, datum_hash: bytes
+) -> None:
+    await confirm_properties(
+        ctx,
+        "confirm_datum_hash",
+        title="Confirm transaction",
+        props=[
+            (
+                "The following transaction output contains datum hash:",
+                format_output_datum_hash(datum_hash),
+            ),
+            ("\nContinue?", None),
+        ],
+        br_code=ButtonRequestType.Other,
+    )
+
+
+async def show_warning_tx_output_misses_datum_hash(ctx: wire.Context) -> None:
```

Comment:
```suggestion
async def show_warning_tx_output_no_datum_hash(ctx: wire.Context) -> None:
```

Reply:
Fixed in https://github.com/vacuumlabs/trezor-firmware/pull/71/commits/b84ee4613aa02f70b3eae2fa53cffa6539aacfe4.

> add a test to sign_tx.plutus.json with a script payment address output and no datum hash

I added such test [here](https://github.com/vacuumlabs/trezor-firmware/pull/71/commits/338a485a34fdc80a0190850424934eb03f8fbf91#diff-1377b9108db8d855501a2e7a09ec9176fa5bc3c6f85cb8ecfbbbe1f902fbcda8R1688), isn't it sufficient?

Reply:
Hm, I don't know. I'd say that since `datum_hash` is connected to plutus functionality it should be rather tested with plutus signing mode. So perhaps we can move/change this test to plutus signing mode. But then we're missing a test for script address output in ORDINARY_TRANSACTION. So I think it's best to have both 😄  No?

Reply:
> since datum_hash is connected to plutus functionality it should be rather tested with plutus signing mode

Well, creating UTXOs with datum hashes is a perfectly valid activity for any signing mode (except pool registrations), I would even argue that creating datum hashes will be more common with ORDINARY signing mode instead of PLUTUS (because of the pattern [funds on a base address -> locked on a script address -> back to a base address]). So I would leave it in ORDINARY tests, but I see no problem in adding similar test case to PLUTUS tests (even though it's a bit redundant).

Reply:
I see, I was again confused about when `datum_hash` should be used. Sorry. Let's keep it as is.